polkadot/runtime/parachains/src/scheduler.rs

```diff
@@ -531,183 +349,160 @@ impl<T: Config> Pallet<T> {
+		// now.
+		for assignment in assignments.rev() {
+			T::AssignmentProvider::push_back_assignment(assignment);
+		}
+	}
 
+	/// For each core that isn't part of the `except_for` set, pop the first item of the claim queue
+	/// and fill the queue from the assignment provider.
+	pub(crate) fn advance_claim_queue(except_for: &BTreeSet<CoreIndex>) {
 		// This can only happen on new sessions at which we move all assignments back to the
 		// provider. Hence, there's nothing we need to do here.
 		if ValidatorGroups::<T>::decode_len().map_or(true, |l| l == 0) {
 			return
 		}
-		let n_session_cores = T::AssignmentProvider::session_core_count();
-		let cq = ClaimQueue::<T>::get();
+		let n_session_cores = Self::num_assigner_cores();
 		let config = configuration::ActiveConfig::<T>::get();
 		// Extra sanity, config should already never be smaller than 1:
 		let n_lookahead = config.scheduler_params.lookahead.max(1);
-		let max_availability_timeouts = config.scheduler_params.max_availability_timeouts;
-		let ttl = config.scheduler_params.ttl;
 
 		for core_idx in 0..n_session_cores {
 			let core_idx = CoreIndex::from(core_idx);
 
-			let n_lookahead_used = cq.get(&core_idx).map_or(0, |v| v.len() as u32);
-
-			// add previously timedout paras back into the queue
-			if let Some(mut entry) = timedout_paras.remove(&core_idx) {
-				if entry.availability_timeouts < max_availability_timeouts {
-					// Increment the timeout counter.
-					entry.availability_timeouts += 1;
-					if n_lookahead_used < n_lookahead {
-						entry.ttl = now + ttl;
-					} else {
-						// Over max capacity, we need to bump ttl (we exceeded the claim queue
-						// size, so otherwise the entry might get dropped before reaching the top):
-						entry.ttl = now + ttl + One::one();
-					}
-					Self::add_to_claim_queue(core_idx, entry);
-					// The claim has been added back into the claim queue.
-					// Do not pop another assignment for the core.
-					continue
-				} else {
-					// Consider timed out assignments for on demand parachains as concluded for
-					// the assignment provider
-					let ret = concluded_paras.insert(core_idx, entry.assignment);
-					debug_assert!(ret.is_none());
+			if !except_for.contains(&core_idx) {
+				let core_idx = CoreIndex::from(core_idx);
+
+				if let Some(dropped_para) = Self::pop_from_claim_queue(&core_idx) {
+					T::AssignmentProvider::report_processed(dropped_para.para_id(), core_idx);
 				}
+
+				Self::fill_claim_queue(core_idx, n_lookahead);
 			}
+		}
+	}
+
+	fn fill_claim_queue(core_idx: CoreIndex, n_lookahead: u32) {
+		ClaimQueue::<T>::mutate(|la| {
+			let cq = la.entry(core_idx).or_default();
 
-			if let Some(concluded_para) = concluded_paras.remove(&core_idx) {
-				T::AssignmentProvider::report_processed(concluded_para);
+			let mut n_lookahead_used = cq.len() as u32;
+
+			// If the claim queue used to be empty, we need to double the first assignment.
```

Comment:
I don't really like this freebie core, but this is the best we can do. We could make the claim queue hold `Option<Assignment>` and this would allow for holes without requiring to clone the assignment, but still, 1 slot would still be wasted.